utils/s3/client.cc

```diff
@@ -847,6 +847,82 @@ data_sink client::make_upload_jumbo_sink(sstring object_name, std::optional<unsi
     return data_sink(std::make_unique<upload_jumbo_sink>(shared_from_this(), std::move(object_name), max_parts_per_piece));
 }
 
+// a sink with known object size
+//
+// unlike upload_sink and upload_jumbo_sink, upload_ranged_sink sends the
+// received buffers right away without accumulating them first. caller is
+// supposed to put() with the payload of part_size until the specified
+// total_size is reached.
+class client::upload_ranged_sink final : public upload_sink_base {
+    // TODO: resume support
+    const size_t _total_size;
+    const size_t _part_size;
+    const unsigned _total_num_parts;
+
+    size_t next_put_size() const {
+        assert(_part_etags.size() <= _total_num_parts);
+        if (_part_etags.size() == _total_num_parts) {
+            return 0;
+        }
+        size_t remainder = _total_size - _part_etags.size() * _part_size;
+        return std::min(remainder, _part_size);
+    }
+
+public:
+    upload_ranged_sink(shared_ptr<client> cln,
+                       sstring object_name,
+                       size_t total_size,
+                       size_t part_size)
+        : upload_sink_base(std::move(cln), std::move(object_name), std::nullopt)
+        , _total_size{total_size}
+        , _part_size{part_size}
+        , _total_num_parts{(total_size + part_size - 1) / part_size} {
+        assert(_part_size >= aws_minimum_part_size);
+        assert(_total_num_parts <= aws_maximum_parts_in_piece);
+    }
+
+    future<> put(temporary_buffer<char> buf) final {
```

Comment:
OK, let's assume we're uploading a 100Gb file. It means that the part size is 100Gb/10000=10Mb. And this sink requires that this 10Mb chunk is submitted here in a single temporary_buffer. That's too restrictive. First, you need to accumulate 10Mb in memory, second, you need to accumulate it in a _contiguous_ buffer

Reply:
i think `output_stream` and its sink are different. please see the test for more details.

Reply:
```
  // so we can test !with_remainder case properly with multiple writes
    assert(total_size % pattern.size() == 0);
    temporary_buffer<char> data(pattern.data(), pattern.size());
    for (size_t bytes_written = 0, buffered = 0;
         bytes_written < total_size;
         bytes_written += pattern.size()) {
        co_await output.write(pattern.data(), pattern.size());
        buffered += pattern.size();
        if (buffered >= part_size) {
            co_await output.flush();
            buffered -= part_size;
        }
    }
```

you put a bunch of temporary_buffer-s of 16-bytes each into output_stream. And the stream accumulates them into 5Mb buffer, then flushes it into sink

But my question above remains unanswered -- what if we upload 100Gb file thus requiring 10Mb of contiguous buffer to be put?

Reply:
thank you Pavel. "contiguous"! i see now. will come up with a better revision.

Reply:
i think the problem of "contiguous" is in the test, not in the implementation. i should pass a `temporary_buffer<char>` to `write()`,  in other words:

```c++
co_await output.write(pattern.data(), pattern.size());
```

is not a real-world use case. while

```c++
co_await output.write(pattern.share());
```
is.

Reply:
> i think the problem of "contiguous" is in the test, not in the implementation

It _is_ in the implementation. Here:

```
future<> put(temporary_buffer<char> buf) final {
    ...
    memory_data_sink_buffers sink_buffers;
    sink_buffers.put(std::move(buf));
    co_await upload_part(std::move(sink_buffers));
```

we see that one buf equals one part. S3 supports at most 10000 parts. 100Gb file means 10Mb parts. 10Mb part means 10Mb tempoarary_buffer which _is_ contiguous

Reply:
@xemul did you get a chance to read https://github.com/scylladb/scylladb/pull/16387#discussion_r1436084985 ? i think it should addresses your concern.

Reply:
Yes, I read that. My concern here is not about using seastar's copy() or any other specific way of configuring this new sink. It's purely about unavoidable use of contiguous large buffer. The net::packet proposal might help, but it's not here and is not extremely nice: quoting from https://github.com/scylladb/seastar/pull/1582#issuecomment-1486750979

> I believe exposing net::packet to higher layers (above the native TCP/IP stack) was a mistake and we should deprecate it with a view towards making it internal.

The new sink should implement `put(temporary_buffer<char> buf)` and it should be a state machine that

* sends `PUT` with `content-length: $part_size` where `$part_size` is _not_ 5Mb as we have currently, but `$file_size / 10000` (with necessary rounding and clamps for small files)
* instantly sends each new `temporary_buffer<>` that is put()-ed into this sink up until the $part_size is fulfilled
* once $part_size is fulfilled -- wrap the part and spawn new `PUT`
* once $file_size is exhausted, pick up parts in flight and send `POST` completing the upload

Reply:
what do you mean by "it's not here"?

Reply:
do you mean we should send a `PUT` request for **each** buffer read from disk?

Reply:
> what do you mean by "it's not here"?

Not in this implementation. The `put(net::packet pkt)` code here also results in part accumulated in memory AFAICS though in non-contiguous form which is definitely better -- it accumulates packets into `_bufs` then sends `upload_part(_bufs.split_at())`

> do you mean we should send a PUT request for each buffer read from disk?

No. I mean that every time you `put(temporary_buffer<> buf)` or `put(net::packet pkt)` the data from buf/pkt should be immediately forwarded to the client's socket via respective part's PUT body http stream. Like ... this

```
    optional<stream<temporary_buffer>> part_body;
    size_t part_uploaded_size;

    future<> put(temporary_buffer buf) {
        if (!part_body.has_value()) {
            part_body.emplace();
            part_uploaded_size = 0;
            (void)[&] () -> future<> {
                    req = request('PUT', headers = { 'content_length': part_size }); // part_size is _large_ i.e. file_size/10`000
                    req.write_body = [&] (output_stream http_body) {
                        buf = co_await body.consume(); // there's not such call in seastar::stream, but I'm too lazy to implement it via existing listen() api. This consume() method should pop up next element from stream and return it
                        co_await part_body.write(buf);
                        part_uploaded_size += buf.size();
                        if (part_uploaded_size == part_size) {
                            // wrap this fiber up somehow
                            part_body.reset();
                            ...
                        }
                    }
                    co_await client->send(req, [] (http::reply rep) {
                        auto etag = rep.get_header('ETag');
                    });
            }(); // in the background
        }
        co_await body->produce(std::move(buf));
    }
```

Reply:
> > what do you mean by "it's not here"?
> 
> Not in this implementation. 

i am confused. it's in the latest revision. 

> The `put(net::packet pkt)` code here also results in part accumulated in memory AFAICS though in non-contiguous form which is definitely better -- it accumulates packets into `_bufs` then sends `upload_part(_bufs.split_at())`
> 
> > do you mean we should send a PUT request for each buffer read from disk?
> 
> No. I mean that every time you `put(temporary_buffer<> buf)` or `put(net::packet pkt)` the data from buf/pkt should be immediately forwarded to the client's socket via respective part's PUT body http stream. Like ... this
> 

the change you are proposing assumes that we only have a single `PUT` request at a given time for uploading the stream of memory chunks. but i think, the right approach is to upload the file with multiple parts in parallel as long as we don't breach the preconfigured limit. if you are fine with the multi-part-in-parallel, i will implement it.

Reply:
@xemul hi Pavel, ping?

Reply:
I still have this concern

> OK, let's assume we're uploading a 100Gb file. It means that the part size is 100Gb/10000=10Mb. And this sink requires that this 10Mb chunk is submitted here in a single temporary_buffer. That's too restrictive. First, you need to accumulate 10Mb in memory, second, you need to accumulate it in a contiguous buffer

The latter restriction (having a single 10Mb buffer) is indeed fixed, but not the former. If you have 100Gb file, then part size is 10Mb at least. So we still need to accumulate 10Mb in RAM. 10Mb is not that much indeed, but 100Gb sstable is not huge either.

Ideally, if you have 100Gb file, you can start to upolad a part as early as the very first buffer of (say) 128k arrives. You send PUT with Content-Length: 10Mb header and those 128k of data. As soon as next 128k arrives, it's sent on a wire as well, and so on until all 80 buffers are sent.